{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "28d29946",
   "metadata": {},
   "source": [
    "### Table of Contents\n",
    "\n",
    "* [Relevant libraries and colour dictionaries](#chapter1)\n",
    "* [SpatialDE](#chapter2)\n",
    "* [Top30](#chapter3)\n",
    "    * [Top 30 compare](#section_3_1)\n",
    "* [Scatterplot](#chapter4)\n",
    "* [AEH](#chapter5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89e12ae7",
   "metadata": {},
   "source": [
    "#### Relevant libraries and colour dictionaries <a class=\"anchor\" id=\"chapter1\"></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c070a404",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import relevant libraries\n",
    "import numpy as np\n",
    "import scanpy as sc\n",
    "import os\n",
    "import pandas as pd\n",
    "import seaborn as sb\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import rcParams\n",
    "from matplotlib import colors\n",
    "import seaborn as sns\n",
    "from collections import OrderedDict\n",
    "from matplotlib import cm\n",
    "import anndata as ann\n",
    "import scanpy.external as sce\n",
    "from datetime import datetime\n",
    "import NaiveDE\n",
    "import SpatialDE\n",
    "from matplotlib_venn import venn3\n",
    "%matplotlib inline\n",
    "\n",
    "# Set current directory\n",
    "os.chdir(\"/Users/mendenlab/work/spatial_granuloma/scripts\")\n",
    "\n",
    "# assign the rigth colours to the right annotation\n",
    "def _set_colors(adata, obs_name, colors):\n",
    "    \"\"\"Set palette with specific colors for specific categories\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    adata : annData\n",
    "    obs_name : column to plot\n",
    "    colors : OrderedDict(): colors named by categories\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "\n",
    "    \"\"\"\n",
    "    if len(colors.values())>0:\n",
    "        palette = []\n",
    "        unique_colors = np.unique(adata.obs[obs_name])\n",
    "        for key in adata.obs[obs_name].cat.categories.tolist():\n",
    "            if key in colors.keys():\n",
    "                palette.append(colors[key])\n",
    "    return palette\n",
    "\n",
    "#Set the colours per annotation\n",
    "spot_colors = []\n",
    "spot_colors = OrderedDict()\n",
    "spot_colors[\"EPIDERMIS\"] = 'blue'\n",
    "spot_colors[\"DERMIS\"] = '#E0EEE0'\n",
    "spot_colors[\"INTERFACE\"] = 'deepskyblue'\n",
    "spot_colors[\"VESSEL\"] = 'darkgreen'\n",
    "spot_colors[\"HAIR FOLLICLE\"] = \"#543005\"\n",
    "spot_colors[\"SWEAT GLAND\"] = 'y'\n",
    "spot_colors[\"SEBACEOUS GLAND\"] = 'mistyrose'\n",
    "spot_colors[\"MUSCLE\"] = 'darkcyan'\n",
    "spot_colors[\"GA\"] = 'firebrick'  \n",
    "spot_colors[\"GNL\"] = 'orchid'\n",
    "spot_colors[\"GSS\"] = 'blueviolet'\n",
    "spot_colors[\"GSC\"] = 'mediumvioletred'\n",
    "spot_colors[\"UNDETERMINED\"] = 'black'\n",
    "\n",
    "\n",
    "dermis_colors = []\n",
    "dermis_colors = OrderedDict()\n",
    "dermis_colors[\"UNDETERMINED\"] = 'black'\n",
    "dermis_colors[\"upper EPIDERMIS\"] = 'blue'\n",
    "dermis_colors[\"middle EPIDERMIS\"] = 'dodgerblue'\n",
    "dermis_colors[\"basal EPIDERMIS\"] = 'skyblue'\n",
    "dermis_colors[\"DERdepth1\"] = '#006837'\n",
    "dermis_colors[\"DERdepth2\"] = '#238443'\n",
    "dermis_colors[\"DERdepth3\"] = '#41AB5D'\n",
    "dermis_colors[\"DERdepth4\"] = '#78C679'\n",
    "dermis_colors[\"DERdepth5\"] = '#ADDD8E'\n",
    "dermis_colors[\"DERdepth6\"] = '#D9F0A3'\n",
    "dermis_colors[\"DERdepth7\"] = '#F7FCB9'\n",
    "\n",
    "leiden_r13_colours = []\n",
    "leiden_r13_colours = OrderedDict()\n",
    "leiden_r13_colours[\"0\"] = 'darkolivegreen'\n",
    "leiden_r13_colours[\"1\"] = \"#D9F0A3\"\n",
    "leiden_r13_colours[\"2\"] = '#238443'\n",
    "leiden_r13_colours[\"3\"] = 'firebrick'\n",
    "leiden_r13_colours[\"4\"] = '#78C679'\n",
    "leiden_r13_colours[\"5\"] = '#78C679'\n",
    "leiden_r13_colours[\"6\"] = '#41AB5D'\n",
    "leiden_r13_colours[\"7\"] = '#006837'\n",
    "leiden_r13_colours[\"8\"] = '#ADDD8E'\n",
    "leiden_r13_colours[\"9\"] = \"#238443\"\n",
    "leiden_r13_colours[\"10\"] = '#78C679'\n",
    "leiden_r13_colours[\"11\"] = 'blue'\n",
    "leiden_r13_colours[\"12\"] = 'orchid'\n",
    "leiden_r13_colours[\"13\"] = '#F46D43'\n",
    "leiden_r13_colours[\"14\"] = 'dodgerblue'\n",
    "leiden_r13_colours[\"15\"] = 'deepskyblue'\n",
    "leiden_r13_colours[\"16\"] = '#cfafaf'\n",
    "leiden_r13_colours[\"17\"] = 'yellow'\n",
    "leiden_r13_colours[\"18\"] = 'darkcyan'\n",
    "leiden_r13_colours[\"19\"] = '#006837'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ab3a7514",
   "metadata": {},
   "outputs": [],
   "source": [
    "def flatten(t):\n",
    "    return [item for sublist in t for item in sublist]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "16bfec5b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n"
     ]
    }
   ],
   "source": [
    "# Import adata \n",
    "adata_path = \"../results/current/\"\n",
    "\n",
    "adata = sc.read(os.path.join(adata_path, \"final/Granuloma_QC_clustering.h5\"))\n",
    "    \n",
    "# setting up \"factors\" with different levels, order = TRUE\n",
    "# add less common annotations LAST so they are not overwritten\n",
    "\n",
    "# Set spot_type and skin_layer as categories and define the levels in each category\n",
    "# Spot type: annatomical annotations\n",
    "adata.obs['spot_type'] = pd.Categorical(\n",
    "    adata.obs['spot_type'],\n",
    "    categories = ['UNDETERMINED', 'DERMIS', \"EPIDERMIS\", 'INTERFACE', 'HAIR FOLLICLE',\n",
    "                'VESSEL', 'MUSCLE', 'SEBACEOUS GLAND', 'SWEAT GLAND', 'GA', 'GNL', 'GSS', 'GSC'],\n",
    "                 ordered = True)\n",
    "\n",
    "# Skin layer\n",
    "adata.obs['skin_layer'] = pd.Categorical(\n",
    "    adata.obs['skin_layer'],\n",
    "    categories = ['UNDETERMINED', \n",
    "                'upper EPIDERMIS', 'middle EPIDERMIS', 'basal EPIDERMIS',\n",
    "                'DERdepth1', 'DERdepth2', 'DERdepth3', 'DERdepth4',\n",
    "                'DERdepth5', 'DERdepth6', 'DERdepth7'],\n",
    "    ordered = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f911e564",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Volumes/Drive/spatial_granuloma/output/SpatialDE\n"
     ]
    }
   ],
   "source": [
    "os.chdir(\"/Volumes/Drive/spatial_granuloma/output/SpatialDE/\") # Set working directory so it saves it in the drive\n",
    "print(os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63baa1d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# os.chdir(\"/Volumes/Drive/spatial_granuloma/output/\") # Set working directory so it saves it in the drive\n",
    "# print('Saving plots in: ' + os.getcwd())\n",
    "# spatialde_path = \"../output/SpatialDE/\"\n",
    "\n",
    "# date = datetime.now().strftime(\"%Y%m%d\") #For timestamp in file names\n",
    "\n",
    "# sample_SPECIMEN_list = adata.obs['sample_SPECIMEN'].unique()\n",
    "\n",
    "# fig, axes = plt.subplots(facecolor='w', edgecolor='k', figsize=(15, 80), dpi=300)\n",
    "# for i, sample_SPECIMEN in enumerate(sample_SPECIMEN_list):\n",
    "#     adata_subset = adata[(adata.obs['sample_SPECIMEN']== sample_SPECIMEN)]\n",
    "#     ax = fig.add_subplot(22, 2, i + 1)\n",
    "    \n",
    "#     if ((adata_subset.obs['sample'].unique() == 'P17851_1002') | (adata_subset.obs['sample'].unique() == 'P18554_1006') | (adata_subset.obs['sample'].unique() == 'P18554_1002')):\n",
    "#         ax.invert_xaxis() \n",
    "#         ax.invert_yaxis()\n",
    "#     plt.scatter(adata_subset.obs['array_col'], \n",
    "#                 -1.5*adata_subset.obs['array_row'],\n",
    "#                 c = adata_subset.obs['spot_type'].map(spot_colors))\n",
    "#     plt.title(sample_SPECIMEN)\n",
    "#     plt.colorbar(ticks=[]);\n",
    "# plt.savefig(spatialde_path + date + '_Separate_slides_colouredbyspottype.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3c8fb54",
   "metadata": {},
   "source": [
    "#### SpatialDE <a class=\"anchor\" id=\"chapter2\"></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5cc1057e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Saving plots in: ../output/SpatialDE/GA_slides/\n",
      "Starting analysis for: P17851_1001_91253-A\n",
      "(1014, 13406)\n",
      "                  array_row  array_col  n_counts\n",
      "AAACGAAGAACATACC          6         64    4625.0\n",
      "AAAGACCCAAGTCGCG         10         48    1079.0\n",
      "AAAGGGATGTAGCAAG         24         62    1040.0\n",
      "AAAGGGCAGCTTGAAT         24         26     673.0\n",
      "AAAGTCACTGATGTAA         10         52     968.0\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/mendenlab/miniconda3/envs/r4.1.2_py3.8/lib/python3.8/site-packages/scipy/optimize/_minpack_py.py:833: OptimizeWarning: Covariance of the parameters could not be estimated\n",
      "  warnings.warn('Covariance of the parameters could not be estimated',\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Input \u001b[0;32mIn [5]\u001b[0m, in \u001b[0;36m<cell line: 13>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     32\u001b[0m counts\u001b[38;5;241m.\u001b[39mto_csv(spatialde_path \u001b[38;5;241m+\u001b[39m date \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m_counts_\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m+\u001b[39m sample_SPECIMEN \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m.csv\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m     33\u001b[0m sc\u001b[38;5;241m.\u001b[39mwrite(os\u001b[38;5;241m.\u001b[39mpath\u001b[38;5;241m.\u001b[39mjoin(spatialde_path, date \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m_adata_subset_\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m+\u001b[39m sample_SPECIMEN \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m.h5\u001b[39m\u001b[38;5;124m'\u001b[39m), adata_subset)\n\u001b[0;32m---> 34\u001b[0m \u001b[43mnorm_expr\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[43mspatialde_path\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[43mdate\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m_norm_expr_\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[43msample_SPECIMEN\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m+\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m.csv\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m     35\u001b[0m resid_expr\u001b[38;5;241m.\u001b[39mto_csv(spatialde_path \u001b[38;5;241m+\u001b[39m date \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m_resid_expr_\u001b[39m\u001b[38;5;124m'\u001b[39m \u001b[38;5;241m+\u001b[39m sample_SPECIMEN \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m.csv\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m     37\u001b[0m \u001b[38;5;66;03m# Run Spatial DEG\u001b[39;00m\n",
      "File \u001b[0;32m~/miniconda3/envs/r4.1.2_py3.8/lib/python3.8/site-packages/pandas/core/generic.py:3563\u001b[0m, in \u001b[0;36mNDFrame.to_csv\u001b[0;34m(self, path_or_buf, sep, na_rep, float_format, columns, header, index, index_label, mode, encoding, compression, quoting, quotechar, line_terminator, chunksize, date_format, doublequote, escapechar, decimal, errors, storage_options)\u001b[0m\n\u001b[1;32m   3552\u001b[0m df \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(\u001b[38;5;28mself\u001b[39m, ABCDataFrame) \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mto_frame()\n\u001b[1;32m   3554\u001b[0m formatter \u001b[38;5;241m=\u001b[39m DataFrameFormatter(\n\u001b[1;32m   3555\u001b[0m     frame\u001b[38;5;241m=\u001b[39mdf,\n\u001b[1;32m   3556\u001b[0m     header\u001b[38;5;241m=\u001b[39mheader,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   3560\u001b[0m     decimal\u001b[38;5;241m=\u001b[39mdecimal,\n\u001b[1;32m   3561\u001b[0m )\n\u001b[0;32m-> 3563\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mDataFrameRenderer\u001b[49m\u001b[43m(\u001b[49m\u001b[43mformatter\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto_csv\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m   3564\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpath_or_buf\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3565\u001b[0m \u001b[43m    \u001b[49m\u001b[43mline_terminator\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mline_terminator\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3566\u001b[0m \u001b[43m    \u001b[49m\u001b[43msep\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msep\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3567\u001b[0m \u001b[43m    \u001b[49m\u001b[43mencoding\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mencoding\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3568\u001b[0m \u001b[43m    \u001b[49m\u001b[43merrors\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43merrors\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3569\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcompression\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcompression\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3570\u001b[0m \u001b[43m    \u001b[49m\u001b[43mquoting\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mquoting\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3571\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3572\u001b[0m \u001b[43m    \u001b[49m\u001b[43mindex_label\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mindex_label\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3573\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmode\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmode\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3574\u001b[0m \u001b[43m    \u001b[49m\u001b[43mchunksize\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunksize\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3575\u001b[0m \u001b[43m    \u001b[49m\u001b[43mquotechar\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mquotechar\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3576\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdate_format\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdate_format\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3577\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdoublequote\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdoublequote\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3578\u001b[0m \u001b[43m    \u001b[49m\u001b[43mescapechar\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mescapechar\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3579\u001b[0m \u001b[43m    \u001b[49m\u001b[43mstorage_options\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mstorage_options\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m   3580\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/r4.1.2_py3.8/lib/python3.8/site-packages/pandas/io/formats/format.py:1180\u001b[0m, in \u001b[0;36mDataFrameRenderer.to_csv\u001b[0;34m(self, path_or_buf, encoding, sep, columns, index_label, mode, compression, quoting, quotechar, line_terminator, chunksize, date_format, doublequote, escapechar, errors, storage_options)\u001b[0m\n\u001b[1;32m   1159\u001b[0m     created_buffer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[1;32m   1161\u001b[0m csv_formatter \u001b[38;5;241m=\u001b[39m CSVFormatter(\n\u001b[1;32m   1162\u001b[0m     path_or_buf\u001b[38;5;241m=\u001b[39mpath_or_buf,\n\u001b[1;32m   1163\u001b[0m     line_terminator\u001b[38;5;241m=\u001b[39mline_terminator,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1178\u001b[0m     formatter\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfmt,\n\u001b[1;32m   1179\u001b[0m )\n\u001b[0;32m-> 1180\u001b[0m \u001b[43mcsv_formatter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msave\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1182\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m created_buffer:\n\u001b[1;32m   1183\u001b[0m     \u001b[38;5;28;01massert\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(path_or_buf, StringIO)\n",
      "File \u001b[0;32m~/miniconda3/envs/r4.1.2_py3.8/lib/python3.8/site-packages/pandas/io/formats/csvs.py:261\u001b[0m, in \u001b[0;36mCSVFormatter.save\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    241\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m get_handle(\n\u001b[1;32m    242\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mfilepath_or_buffer,\n\u001b[1;32m    243\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mmode,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    249\u001b[0m \n\u001b[1;32m    250\u001b[0m     \u001b[38;5;66;03m# Note: self.encoding is irrelevant here\u001b[39;00m\n\u001b[1;32m    251\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mwriter \u001b[38;5;241m=\u001b[39m csvlib\u001b[38;5;241m.\u001b[39mwriter(\n\u001b[1;32m    252\u001b[0m         handles\u001b[38;5;241m.\u001b[39mhandle,\n\u001b[1;32m    253\u001b[0m         lineterminator\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mline_terminator,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    258\u001b[0m         quotechar\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mquotechar,\n\u001b[1;32m    259\u001b[0m     )\n\u001b[0;32m--> 261\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_save\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/r4.1.2_py3.8/lib/python3.8/site-packages/pandas/io/formats/csvs.py:266\u001b[0m, in \u001b[0;36mCSVFormatter._save\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    264\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_need_to_save_header:\n\u001b[1;32m    265\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_save_header()\n\u001b[0;32m--> 266\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_save_body\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/r4.1.2_py3.8/lib/python3.8/site-packages/pandas/io/formats/csvs.py:304\u001b[0m, in \u001b[0;36mCSVFormatter._save_body\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    302\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m start_i \u001b[38;5;241m>\u001b[39m\u001b[38;5;241m=\u001b[39m end_i:\n\u001b[1;32m    303\u001b[0m     \u001b[38;5;28;01mbreak\u001b[39;00m\n\u001b[0;32m--> 304\u001b[0m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_save_chunk\u001b[49m\u001b[43m(\u001b[49m\u001b[43mstart_i\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mend_i\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/r4.1.2_py3.8/lib/python3.8/site-packages/pandas/io/formats/csvs.py:315\u001b[0m, in \u001b[0;36mCSVFormatter._save_chunk\u001b[0;34m(self, start_i, end_i)\u001b[0m\n\u001b[1;32m    312\u001b[0m data \u001b[38;5;241m=\u001b[39m [res\u001b[38;5;241m.\u001b[39miget_values(i) \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;28mlen\u001b[39m(res\u001b[38;5;241m.\u001b[39mitems))]\n\u001b[1;32m    314\u001b[0m ix \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdata_index[slicer]\u001b[38;5;241m.\u001b[39m_format_native_types(\u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_number_format)\n\u001b[0;32m--> 315\u001b[0m \u001b[43mlibwriters\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwrite_csv_rows\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    316\u001b[0m \u001b[43m    \u001b[49m\u001b[43mdata\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    317\u001b[0m \u001b[43m    \u001b[49m\u001b[43mix\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    318\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mnlevels\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    319\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcols\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    320\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwriter\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    321\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/r4.1.2_py3.8/lib/python3.8/site-packages/pandas/_libs/writers.pyx:75\u001b[0m, in \u001b[0;36mpandas._libs.writers.write_csv_rows\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "os.chdir(\"/Volumes/Drive/spatial_granuloma/output/\") # Set working directory so it saves it in the drive\n",
    "spatialde_path = \"../output/SpatialDE/GA_slides/\"\n",
    "date = datetime.now().strftime(\"%Y%m%d\") #For timestamp in file names\n",
    "print('Saving plots in: ' + spatialde_path)\n",
    "\n",
    "# Subset by sample_SPECIMEN, then select the subsets we actually want to check - with GS or NL spots\n",
    "sample_SPECIMEN_list = ['P17851_1001_91253-A', 'P17851_1002_91253-A', 'P17851_1002_91253-B', \n",
    "                        'P17851_1003_45703-A', 'P17851_1004_45703-A', 'P18554_1001_50107-A',\n",
    "                        'P18554_1002_50107-A']\n",
    "timestamp = '20220511' # timestamp of saved files, check if they are the most recent in folder\n",
    "# Loop\n",
    "# -------------------------------------------------------------------------------------------------------\n",
    "for sample_SPECIMEN in sample_SPECIMEN_list:\n",
    "    \n",
    "    print('Starting analysis for: ' + sample_SPECIMEN)\n",
    "    adata_subset = adata[(adata.obs['sample_SPECIMEN']== sample_SPECIMEN)].copy()\n",
    "    \n",
    "    # Get the counts dataframe\n",
    "    counts = adata_subset.to_df(layer=\"counts\")\n",
    "    counts = counts.T[counts.sum(0) >= 3].T  # Filter practically unobserved genes\n",
    "    print(counts.shape)\n",
    "    \n",
    "    # Correct for library size or sequencing depth\n",
    "    sample_info = sc.get.obs_df(adata_subset, keys=['array_row', 'array_col', \"n_counts\"])\n",
    "    counts = counts.loc[sample_info.index]  # Align count matrix with metadata table\n",
    "    print(sample_info.head(5))\n",
    "    norm_expr = NaiveDE.stabilize(counts.T).T\n",
    "    resid_expr = NaiveDE.regress_out(sample_info, norm_expr.T, 'np.log(n_counts)').T\n",
    "    \n",
    "    # Save all the data\n",
    "    sample_info.to_csv(spatialde_path + date + '_sample_info_' + sample_SPECIMEN + '.csv')\n",
    "    counts.to_csv(spatialde_path + date + '_counts_' + sample_SPECIMEN + '.csv')\n",
    "    sc.write(os.path.join(spatialde_path, date + '_adata_subset_' + sample_SPECIMEN + '.h5'), adata_subset)\n",
    "    norm_expr.to_csv(spatialde_path + date + '_norm_expr_' + sample_SPECIMEN + '.csv')\n",
    "    resid_expr.to_csv(spatialde_path + date + '_resid_expr_' + sample_SPECIMEN + '.csv')\n",
    "    \n",
    "    # Run Spatial DEG\n",
    "    print('Saving plot as: ' + spatialde_path + date + sample_SPECIMEN + '.csv')\n",
    "    if not os.path.isfile(os.path.join(spatialde_path + timestamp + '_results_' + sample_SPECIMEN + '.csv')):\n",
    "        print('Calculating spatial DE for patient 1 slide 1')\n",
    "        X = sample_info[['array_row', 'array_col']]\n",
    "        results = SpatialDE.run(X, resid_expr)\n",
    "        results.to_csv(spatialde_path + date + '_results_' + sample_SPECIMEN + '.csv')\n",
    "        print('Wohoo!' + sample_SPECIMEN + ' data saved!')\n",
    "        print(\"\")\n",
    "    \n",
    "    # or load it if already computed\n",
    "    else:\n",
    "        print('Spatial DE already computed for ' + sample_SPECIMEN + ', importing results . csv file from: ' + spatialde_path)\n",
    "        print('Need to load one by one, or write new code to load them all!:)')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e716a29",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visualise spatial slides\n",
    "for sample_SPECIMEN in sample_SPECIMEN_list:\n",
    "    adata_subset = adata[(adata.obs['sample_SPECIMEN']== sample_SPECIMEN)].copy()\n",
    "    fig,axes = plt.subplots(facecolor = 'w', edgecolor = 'k', figsize = (40, 20))\n",
    "    \n",
    "    ax = plt.subplot(3, 3, 1)\n",
    "    if ((adata_subset.obs['sample'].unique() == 'P17851_1002') | (adata_subset.obs['sample'].unique() == 'P18554_1006') | (adata_subset.obs['sample'].unique() == 'P18554_1002')):\n",
    "        ax.invert_xaxis() \n",
    "        ax.invert_yaxis()\n",
    "    plt.title('coloured by spot type', fontsize = 20)\n",
    "    plt.axis('equal')\n",
    "    plt.scatter(adata_subset.obs['array_col'], -1.5*adata_subset.obs['array_row'], c = adata_subset.obs['spot_type'].map(spot_colors))\n",
    "    \n",
    "    ax = plt.subplot(3, 3, 2)\n",
    "    if ((adata_subset.obs['sample'].unique() == 'P17851_1002') | (adata_subset.obs['sample'].unique() == 'P18554_1006') | (adata_subset.obs['sample'].unique() == 'P18554_1002')):\n",
    "        ax.invert_xaxis() \n",
    "        ax.invert_yaxis()\n",
    "    plt.title(sample_SPECIMEN + '\\n coloured by skin layer', fontsize = 20)\n",
    "    plt.axis('equal')\n",
    "    plt.scatter(adata_subset.obs['array_col'], -1.5*adata_subset.obs['array_row'], c = adata_subset.obs['skin_layer'].map(dermis_colors))\n",
    "    \n",
    "    ax = plt.subplot(3, 3, 3)\n",
    "    if ((adata_subset.obs['sample'].unique() == 'P17851_1002') | (adata_subset.obs['sample'].unique() == 'P18554_1006') | (adata_subset.obs['sample'].unique() == 'P18554_1002')):\n",
    "        ax.invert_xaxis() \n",
    "        ax.invert_yaxis()\n",
    "    plt.axis('equal')\n",
    "    plt.title('coloured by leiden clusters (r = 1.3)', fontsize = 20)\n",
    "    plt.scatter(adata_subset.obs['array_col'], -1.5*adata_subset.obs['array_row'], c = adata_subset.obs['leiden_r1.3_patient'].map(leiden_r13_colours))\n",
    "    plt.axis('equal')\n",
    "    plt.colorbar(ticks=[]);\n",
    "    plt.savefig(spatialde_path + date + '_Slides_' + sample_SPECIMEN + '.png', bbox_inches='tight', dpi = 120)\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b98ba738",
   "metadata": {},
   "source": [
    "#### Top 30 spatially differentially expressed genes <a class=\"anchor\" id=\"chapter3\"></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c27a5281",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"/Volumes/Drive/spatial_granuloma/output/\") # Set working directory so it saves it in the drive\n",
    "spatialde_path = \"../output/SpatialDE/GA_slides/\"\n",
    "date = datetime.now().strftime(\"%Y%m%d\") #For timestamp in file names\n",
    "print('Saving plots in: ' + spatialde_path)\n",
    "\n",
    "# Subset by sample_SPECIMEN, then select the subsets we actually want to check - with GS or NL spots\n",
    "sample_SPECIMEN_list = ['P17851_1001_91253-A', 'P17851_1002_91253-A', 'P17851_1002_91253-B', \n",
    "                        'P17851_1003_45703-A', 'P17851_1004_45703-A', 'P18554_1001_50107-A',\n",
    "                        'P18554_1002_50107-A']\n",
    "\n",
    "timestamp = '20220511' # the timestamp of the files (check filename)\n",
    "\n",
    "# Loop\n",
    "# -------------------------------------------------------------------------------------------------------\n",
    "\n",
    "for sample_SPECIMEN in sample_SPECIMEN_list:\n",
    "    print('Starting analysis for: ' + sample_SPECIMEN) \n",
    "    # Subset adata \n",
    "    adata_sample_SPECIMEN = adata[(adata.obs['sample_SPECIMEN']== sample_SPECIMEN)].copy()\n",
    "    \n",
    "    # Load the data for the particular sample\n",
    "    results = pd.read_csv(spatialde_path + timestamp + '_results_' + sample_SPECIMEN + '.csv')\n",
    "    sample_info = pd.read_csv(spatialde_path + timestamp + '_sample_info_' + sample_SPECIMEN + '.csv')\n",
    "    counts = pd.read_csv(spatialde_path + timestamp + '_counts_' + sample_SPECIMEN + '.csv')\n",
    "    norm_expr = pd.read_csv(spatialde_path + timestamp + '_norm_expr_' + sample_SPECIMEN + '.csv')\n",
    "    resid_expr = pd.read_csv(spatialde_path + timestamp + '_resid_expr_' + sample_SPECIMEN + '.csv')\n",
    "    \n",
    "    # Print results\n",
    "    print(results.sort_values('qval').head(10)[['g', 'l', 'qval']])\n",
    "\n",
    "    # Genes - top 30 by qval\n",
    "    genes = results.sort_values('qval').head(30)['g'].tolist()\n",
    "    textfile = open(spatialde_path + 'Top30/' + date + '_top30genes_' + sample_SPECIMEN + '.txt', \"w\")\n",
    "    for gene in genes:\n",
    "        textfile.write(gene + \"\\n\")\n",
    "    textfile.close()\n",
    "    \n",
    "    # Plot genes in spatial slides - top 30 by qval\n",
    "    print('Plotting genes in spatial slides!')\n",
    "    fig, axes = plt.subplots(facecolor='w', edgecolor='k', figsize=(20, 80))\n",
    "    for i, g in enumerate(genes):\n",
    "        ax = plt.subplot(15, 2, 1)\n",
    "        if ((adata_sample_SPECIMEN.obs['sample'].unique() == 'P17851_1002') | (adata_sample_SPECIMEN.obs['sample'].unique() == 'P18554_1006') | (adata_sample_SPECIMEN.obs['sample'].unique() == 'P18554_1002')):\n",
    "            ax.invert_xaxis() \n",
    "            ax.invert_yaxis()\n",
    "        plt.scatter(adata_sample_SPECIMEN.obs['array_col'], -1.5*adata_sample_SPECIMEN.obs['array_row'], c = adata_sample_SPECIMEN.obs['spot_type'].map(spot_colors))\n",
    "        plt.axis('equal')\n",
    "        \n",
    "        ax = plt.subplot(15, 2, i + 1)\n",
    "        if ((adata_sample_SPECIMEN.obs['sample'].unique() == 'P17851_1002') | (adata_sample_SPECIMEN.obs['sample'].unique() == 'P18554_1006') | (adata_sample_SPECIMEN.obs['sample'].unique() == 'P18554_1002')):\n",
    "            ax.invert_xaxis() \n",
    "            ax.invert_yaxis()\n",
    "        plt.scatter(sample_info['array_col'], -1.5*sample_info['array_row'], c=norm_expr[g], s = 80);\n",
    "        plt.title(g, fontsize=20)\n",
    "        plt.axis('equal')\n",
    "        plt.colorbar(ticks=[]);\n",
    "    plt.savefig(spatialde_path + 'Top30/' + date + '_Top30qval_' + sample_SPECIMEN + '.png', bbox_inches='tight', dpi = 120)\n",
    "    plt.show()\n",
    "    print('Wohoo! Plots saved for: ' + sample_SPECIMEN)\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6debfb04",
   "metadata": {},
   "source": [
    "#### Top 30 compare  <a class=\"anchor\" id=\"section_3_1\"></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e359167",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"/Volumes/Drive/spatial_granuloma/output/\") # Set working directory so it saves it in the drive\n",
    "spatialde_path = \"../output/SpatialDE/GA_slides/\"\n",
    "date = datetime.now().strftime(\"%Y%m%d\") #For timestamp in file names\n",
    "print('Saving plots in: ' + spatialde_path)\n",
    "\n",
    "# Subset by sample_SPECIMEN, then select the subsets we actually want to check - with GS or NL spots\n",
    "sample_SPECIMEN_list = ['P17851_1001_91253-A', 'P17851_1002_91253-A', 'P17851_1002_91253-B', \n",
    "                        'P17851_1003_45703-A', 'P17851_1004_45703-A', 'P18554_1001_50107-A',\n",
    "                        'P18554_1002_50107-A']\n",
    "\n",
    "timestamp = '20220511' # the timestamp of the files (check filename)\n",
    "\n",
    "# Loop to read genes .txt file\n",
    "# -------------------------------------------------------------------------------------------------------\n",
    "all_top30genes = {}\n",
    "for sample_SPECIMEN in sample_SPECIMEN_list:\n",
    "    file_path = spatialde_path + 'Top30/' + timestamp + '_top30genes_' + sample_SPECIMEN + '.txt'\n",
    "    \n",
    "    with open(file_path) as f:\n",
    "        all_top30genes[sample_SPECIMEN] = f.read().splitlines()\n",
    "    \n",
    "#print(all_top30genes)\n",
    "df = pd.DataFrame(all_top30genes.items(), columns = ['sample_SPECIMEN', 'genes']).explode('genes')\n",
    "df.genes.value_counts() > 1\n",
    "v = df.genes.value_counts()\n",
    "df_filtered = df[df.genes.isin(v.index[v.gt(1)])]\n",
    "#df['genes'].value_counts().plot(kind='bar')\n",
    "\n",
    "# Plot common genes\n",
    "# -------------------------------------------------------------------------------------------------------\n",
    "df_plot = df_filtered.groupby(['sample_SPECIMEN', 'genes']).size().reset_index().pivot(columns='sample_SPECIMEN', index='genes', values=0)\n",
    "df_plot\n",
    "\n",
    "plt.figure(facecolor='w', edgecolor='k', figsize=(60, 60))\n",
    "df_plot.plot.bar(stacked = True, figsize=(15, 8),\n",
    "                 color = ['dodgerblue', 'deepskyblue', 'lightblue', 'orange', 'gold', 'purple', 'fuchsia', '#8FBC8F', '#C1FFC1', '#8B0000', '#FF0000'])\n",
    "plt.title(\"Common top 30 spatially-DE genes \\n in granuloma spots (GA)\")\n",
    "plt.xlabel(\"Common genes\")\n",
    "plt.ylabel(\"Count\")\n",
    "plt.legend(loc='center left', bbox_to_anchor=(1.0, 0.75))\n",
    "plt.savefig(spatialde_path + 'Top30/' + date + '_Barplot_common_top30' + '.png', bbox_inches='tight', dpi = 120)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d74f8cc",
   "metadata": {},
   "source": [
    "#### Scatterplots <a class=\"anchor\" id=\"chapter4\"></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "104f9a42",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"/Volumes/Drive/spatial_granuloma/output/\") # Set working directory so it saves it in the drive\n",
    "spatialde_path = \"../output/SpatialDE/GA_slides/\"\n",
    "date = datetime.now().strftime(\"%Y%m%d\") #For timestamp in file names\n",
    "print('Saving plots in: ' + spatialde_path)\n",
    "\n",
    "# Subset by sample_SPECIMEN, then select the subsets we actually want to check - with GS or NL spots\n",
    "sample_SPECIMEN_list = ['P17851_1001_91253-A', 'P17851_1002_91253-A', 'P17851_1002_91253-B', \n",
    "                        'P17851_1003_45703-A', 'P17851_1004_45703-A', 'P18554_1001_50107-A',\n",
    "                        'P18554_1002_50107-A']\n",
    "\n",
    "timestamp = '20220511' # the timestamp of the files (check filename)\n",
    "\n",
    "# Loop\n",
    "# -------------------------------------------------------------------------------------------------------\n",
    "\n",
    "for sample_SPECIMEN in sample_SPECIMEN_list:\n",
    "    print('Starting analysis for: ' + sample_SPECIMEN) \n",
    "    # Subset adata \n",
    "    adata_sample_SPECIMEN = adata[(adata.obs['sample_SPECIMEN']== sample_SPECIMEN)].copy()\n",
    "\n",
    "    # Load the data for the particular sample\n",
    "    results = pd.read_csv(spatialde_path + timestamp + '_results_' + sample_SPECIMEN + '.csv')\n",
    "    norm_expr = pd.read_csv(spatialde_path + timestamp + '_norm_expr_' + sample_SPECIMEN + '.csv')\n",
    "#     sample_info = pd.read_csv(spatialde_path + 'SpatialDE_Zoom/' + timestamp + '_sample_info_' + sample_SPECIMEN + '.csv')\n",
    "#     counts = pd.read_csv(spatialde_path + 'SpatialDE_Zoom/' + timestamp + '_counts_' + sample_SPECIMEN + '.csv')\n",
    "#     resid_expr = pd.read_csv(spatialde_path + 'SpatialDE_Zoom/' + timestamp + '_resid_expr_' + sample_SPECIMEN + '.csv')\n",
    "    \n",
    "    # Highlight gene groups\n",
    "    group1 = results.sort_values('FSV').tail(3)[['g', 'l', 'qval', 'FSV']] # highest FSV\n",
    "    print(group1)\n",
    "    group2 = results.sort_values('qval').head(3)[['g', 'l', 'qval', 'FSV']] # lowest q-value\n",
    "    print(group2)    \n",
    "    group3 = results[(results.qval > 1e-10) & (results.FSV < 0.98)].sort_values('FSV').tail(3)[['g', 'l', 'qval', 'FSV']]\n",
    "    print(group3)\n",
    "    group4 = results[(results.qval > 1e-15) & (results.FSV < 0.9)].sort_values(['qval', 'FSV'], ascending = [True, False]).head(3)[['g', 'l', 'qval', 'FSV']]\n",
    "    print(group4)\n",
    "    group5 = results[(results.qval < 1e-2) & (results.qval > 1e-4) & (results.FSV < 0.9)].sort_values(['FSV', 'qval'], ascending = [False, True]).head(3)[['g', 'l', 'qval', 'FSV']]\n",
    "    print(group5)\n",
    "    \n",
    "    # Plot scatterplot\n",
    "    print('Plotting scatterplot!')\n",
    "    fig = plt.figure(facecolor='w', edgecolor='k', figsize=(10, 5))\n",
    "    plt.yscale('log')\n",
    "    plt.title(sample_SPECIMEN)\n",
    "    plt.scatter(results['FSV'], \n",
    "                results['qval'] + 1e-20, \n",
    "                c = 'lightgray')\n",
    "#     plt.scatter(results.loc[results['g'].isin(group1.g)]['FSV'], \n",
    "#                 results.loc[results['g'].isin(group1.g)]['qval'] + 1e-20, # we have to sum 1e-18 to solve the problem of log(0)\n",
    "#                 c = 'red')\n",
    "#     plt.scatter(results.loc[results['g'].isin(group2.g)]['FSV'], \n",
    "#                 results.loc[results['g'].isin(group2.g)]['qval'] + 1e-20, \n",
    "#                 c = 'purple')\n",
    "#     plt.scatter(results.loc[results['g'].isin(group3.g)]['FSV'], \n",
    "#                 results.loc[results['g'].isin(group3.g)]['qval'] + 1e-20, \n",
    "#                 c = 'green')\n",
    "#     plt.scatter(results.loc[results['g'].isin(group4.g)]['FSV'], \n",
    "#                 results.loc[results['g'].isin(group4.g)]['qval'] + 1e-20, \n",
    "#                 c = 'orange')\n",
    "#     plt.scatter(results.loc[results['g'].isin(group5.g)]['FSV'], \n",
    "#                 results.loc[results['g'].isin(group5.g)]['qval'] + 1e-20, \n",
    "#                 c = 'blue')\n",
    "\n",
    "    plt.axhline(0.05, c='black', lw=1, ls='--')\n",
    "    plt.axhline(0.01, c='black', lw=1, ls='--')\n",
    "    \n",
    "    plt.gca().invert_yaxis();\n",
    "    plt.xlabel('Fraction spatial variance')\n",
    "    plt.ylabel('Adj. P-value');\n",
    "    plt.savefig(spatialde_path + 'Scatterplots/' + date + '_Scatterplot_' + sample_SPECIMEN + '.png')\n",
    "    plt.show()\n",
    "    \n",
    "    print('Wohoo! Plots saved for: ' + sample_SPECIMEN)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbd76eba",
   "metadata": {},
   "outputs": [],
   "source": [
    "patterns.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b406df2",
   "metadata": {},
   "outputs": [],
   "source": [
    "patterns.head()\n",
    "plt.scatter(95*sample_info['array_col'], \n",
    "            -1.5*sample_info['array_row'],\n",
    "            c = patterns[0])\n",
    "plt.colorbar(ticks=[]);"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a70cf259",
   "metadata": {},
   "source": [
    "#### AEH <a class=\"anchor\" id=\"chapter5\"></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "78d9121e",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"/Volumes/Drive/spatial_granuloma/output/\") # Set working directory so it saves it in the drive\n",
    "spatialde_path = \"../output/SpatialDE/GA_slides/\"\n",
    "date = datetime.now().strftime(\"%Y%m%d\") #For timestamp in file names\n",
    "print('Saving plots in: ' + spatialde_path)\n",
    "\n",
    "# Subset by sample_SPECIMEN, then select the subsets we actually want to check - with GA spots\n",
    "sample_SPECIMEN_list = ['P17851_1001_91253-A', 'P17851_1002_91253-A', 'P17851_1002_91253-B', \n",
    "                        'P17851_1003_45703-A', 'P17851_1004_45703-A', 'P18554_1001_50107-A',\n",
    "                        'P18554_1002_50107-A']\n",
    "\n",
    "timestamp = '20220511' # the timestamp of the files (check filename)\n",
    "lengthscale_vector = [1]\n",
    "conditions = ['qval < 0.01']\n",
    "number_of_patterns = 8\n",
    "\n",
    "# Loop for each sample_SPECIMEN\n",
    "# -------------------------------------------------------------------------------------------------------\n",
    "\n",
    "plot_only = True # If false then we will compute AEH from scratch \n",
    "\n",
    "for sample_SPECIMEN in sample_SPECIMEN_list:\n",
    "    print('Starting analysis for: ' + sample_SPECIMEN) \n",
    "    # Subset adata \n",
    "    adata_sample_SPECIMEN = adata[(adata.obs['sample_SPECIMEN']== sample_SPECIMEN)].copy()\n",
    "\n",
    "    # Load the data for the particular sample\n",
    "    results = pd.read_csv(spatialde_path + timestamp + '_results_' + sample_SPECIMEN + '.csv')\n",
    "    norm_expr = pd.read_csv(spatialde_path + timestamp + '_norm_expr_' + sample_SPECIMEN + '.csv')\n",
    "    sample_info = pd.read_csv(spatialde_path + timestamp + '_sample_info_' + sample_SPECIMEN + '.csv')\n",
    "    counts = pd.read_csv(spatialde_path + timestamp + '_counts_' + sample_SPECIMEN + '.csv')\n",
    "    resid_expr = pd.read_csv(spatialde_path + timestamp + '_resid_expr_' + sample_SPECIMEN + '.csv')\n",
    "    \n",
    "    X = sample_info[['array_row', 'array_col']]\n",
    "    \n",
    "    # Loop for AEH\n",
    "    for cond in conditions:\n",
    "        print('Subsetting results for condition: ' + cond)\n",
    "        print('')\n",
    "        sign_results = results.query(cond)\n",
    "        \n",
    "        if plot_only == False:\n",
    "            print('Saving df as: ' + spatialde_path + 'AEH/' + date + '_sign_results_' + cond + '.csv')\n",
    "            sign_results.to_csv(spatialde_path + 'AEH/' + date + '_' + sample_SPECIMEN + '_sign_results_' + cond + '.csv')\n",
    "\n",
    "        for lengthscale in lengthscale_vector:\n",
    "            if (plot_only == False):\n",
    "                print('calculating spatial de for lengthscale = ' + str(lengthscale))\n",
    "                print('')\n",
    "                histology_results, patterns = SpatialDE.aeh.spatial_patterns(X, resid_expr, sign_results, C=number_of_patterns, \n",
    "                                                                             l=lengthscale, verbosity=1)\n",
    "                print('Saving spatial DE results')\n",
    "                print('')\n",
    "                patterns.to_csv(spatialde_path + 'AEH/' + date + '_' + sample_SPECIMEN + '_patterns_' + cond + '_lengthscale' + str(lengthscale) + '.csv')\n",
    "                histology_results.to_csv(spatialde_path + 'AEH/' + date + '_' +  sample_SPECIMEN + '_histology_results_' + cond + '_lengthscale' + str(lengthscale) + '.csv')\n",
    "            else:\n",
    "                patterns = pd.read_csv(spatialde_path + 'AEH/' + timestamp + '_' + sample_SPECIMEN + '_patterns_' + cond + '_lengthscale' + str(lengthscale) + '.csv', index_col=0)\n",
    "                histology_results = pd.read_csv(spatialde_path + 'AEH/' + timestamp + '_' +  sample_SPECIMEN + '_histology_results_' + cond + '_lengthscale' + str(lengthscale) + '.csv', index_col=0)\n",
    "            \n",
    "            print('Computing plots')\n",
    "            print('')\n",
    "            fig,axes = plt.subplots(facecolor='w', edgecolor='k', figsize=(20, int(3*number_of_patterns)))\n",
    "            for i in range(number_of_patterns):\n",
    "                fig.suptitle(cond + '_lengthscale' + str(lengthscale), fontsize=22, y = 0.9)\n",
    "                ax = plt.subplot(int(number_of_patterns/2), 2, i + 1)\n",
    "                if ((adata_sample_SPECIMEN.obs['sample'].unique() == 'P17851_1002') | (adata_sample_SPECIMEN.obs['sample'].unique() == 'P18554_1006') | (adata_sample_SPECIMEN.obs['sample'].unique() == 'P18554_1002')):\n",
    "                    ax.invert_xaxis() \n",
    "                    ax.invert_yaxis()\n",
    "                plt.scatter(sample_info['array_col'], -1.5*sample_info['array_row'], c = patterns[str(i)], s = 80); # Note: if patterns.csv is read in, you might need to change it to patterns[str(i)] \n",
    "                plt.axis('equal')\n",
    "                plt.title('Pattern {} - {} genes'.format(i, histology_results.query('pattern == @i').shape[0] ))\n",
    "                plt.colorbar(ticks=[]);\n",
    "                plt.savefig(spatialde_path + 'AEH/' + date + '_' + sample_SPECIMEN + '_patterns_' + cond + '_lengthscale' + str(lengthscale) + '.png', bbox_inches='tight', dpi = 120)\n",
    "            print('Patterns successfully plotted and saved!')\n",
    "            print('')\n",
    "            \n",
    "    print('Wohoo! AEH finished for: ' + sample_SPECIMEN + 'condition ' + cond)\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Rpy2 kernel",
   "language": "python",
   "name": "r4.1.2_py3.8"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
